File: julia/problem_18.jl
# Copyright 2014 Mitchell Kember. Subject to the MIT License.
# Project Euler: Problem 18
# Maximum path sum I

module problem_18

const mat = [
	75 64 82 10 65 34 67 92 33 29 14 57 48 31 23;
	95 47 87 47 03 63 70 70 94 51 17 29 40 04  0;
	17 35 82 75 07 16 80 16 97 68 27 87 60  0  0;
	18 04 23 73 06 40 37 52 39 50 69 53  0  0  0;
	20 01 77 28 83 32 91 78 58 16 38  0  0  0  0;
	19 02 04 56 47 43 17 43 73 93  0  0  0  0  0;
	88 65 26 33 25 73 91 30 73  0  0  0  0  0  0;
	99 41 72 65 77 14 67 98  0  0  0  0  0  0  0;
	41 48 44 28 17 53 70  0  0  0  0  0  0  0  0;
	41 71 33 38 89 09  0  0  0  0  0  0  0  0  0;
	53 11 52 68 23  0  0  0  0  0  0  0  0  0  0;
	70 71 04 27  0  0  0  0  0  0  0  0  0  0  0;
	91 66 98  0  0  0  0  0  0  0  0  0  0  0  0;
	63 62  0  0  0  0  0  0  0  0  0  0  0  0  0;
	04  0  0  0  0  0  0  0  0  0  0  0  0  0  0;
]

function maxsum(i, j)
	x = mat[i,j]
	if i + j == size(mat, 1) + 1
		return x
	end
	x + max(maxsum(i+1, j), maxsum(i, j+1))
end

function solve()
	maxsum(1, 1)
end

end # module

File: c++/problem_18.cpp
// Copyright 2016 Mitchell Kember. Subject to the MIT License.
// Project Euler: Problem 18
// Maximum path sum I

#include "common.hpp"

namespace problem_18 {

constexpr long side_length = 15;
constexpr long triangle[side_length][side_length] = {
	{75},
	{95, 64},
	{17, 47, 82},
	{18, 35, 87, 10},
	{20,  4, 82, 47, 65},
	{19,  1, 23, 75,  3, 34},
	{88,  2, 77, 73,  7, 63, 67},
	{99, 65,  4, 28,  6, 16, 70, 92},
	{41, 41, 26, 56, 83, 40, 80, 70, 33},
	{41, 48, 72, 33, 47, 32, 37, 16, 94, 29},
	{53, 71, 44, 65, 25, 43, 91, 52, 97, 51, 14},
	{70, 11, 33, 28, 77, 73, 17, 78, 39, 68, 17, 57},
	{91, 71, 52, 38, 17, 14, 91, 43, 58, 50, 27, 29, 48},
	{63, 66,  4, 68, 89, 53, 67, 30, 73, 16, 69, 87, 40, 31},
	{ 4, 62, 98, 27, 23,  9, 70, 98, 73, 93, 38, 53, 60,  4, 23}
};

long max_path_sum(long total, long i, long j) {
	if (i >= side_length) {
		return total;
	}
	total += triangle[i][j];
	const long left = max_path_sum(total, i+1, j);
	const long right = max_path_sum(total, i+1, j+1);
	return common::max(left, right);
}

long solve() {
	return max_path_sum(0, 0, 0);
}

} // namespace problem_18
